src/Data/Singletons/Single/Fixity.hs

```diff
@@ -0,0 +1,27 @@
+module Data.Singletons.Single.Fixity where
+
+import Prelude hiding ( exp )
+import Language.Haskell.TH hiding ( cxt )
+import Language.Haskell.TH.Syntax (Quasi(..))
+import Data.Singletons.Util
+import Data.Singletons.Names
+import Language.Haskell.TH.Desugar
+
+singInfixDecl :: Fixity -> Name -> DLetDec
+singInfixDecl fixity name
+  | isUpcase name =
+    -- is it a tycon name or a datacon name??
+    -- it *must* be a datacon name, because symbolic tycons
+    -- can't be promoted. This is terrible.
```

Comment:
This comment is out-of-date, because `TypeInType` allows you to promote symbolic tycons and datacons. Fixing this would take Some Work, unfortunately, because you'd have to poke around to see what's in scope. In any case, it's orthogonal to your patch.